v2/cli/secrets_commands.go

```diff
@@ -0,0 +1,224 @@
+		},
+		{
+			Name:  "set",
+			Usage: "Define or redefine the value of one or more secrets",
+			Flags: []cli.Flag{
+				&cli.StringFlag{
+					Name:     flagProject,
+					Aliases:  []string{"p"},
+					Usage:    "Set secrets for the specified project (required)",
+					Required: true,
+				},
+				&cli.StringSliceFlag{
+					Name:    flagSet,
+					Aliases: []string{"s"},
+					Usage: "Set a secret using the specified key=value pair " +
+						"(required)",
+					Required: true,
+				},
+			},
+			Action: secretsSet,
+		},
+		{
+			Name:  "unset",
+			Usage: "Clear the value of one or more secrets",
+			Flags: []cli.Flag{
+				&cli.StringFlag{
+					Name:     flagProject,
+					Aliases:  []string{"p"},
+					Usage:    "Clear secrets for the specified project",
+					Required: true,
+				},
+				&cli.StringSliceFlag{
+					Name:     flagUnset,
+					Aliases:  []string{"u"},
+					Usage:    "Clear a secret haveing the specified key (required)",
+					Required: true,
+				},
+			},
+			Action: secretsUnset,
+		},
+	},
+}
+
+func secretsList(c *cli.Context) error {
+	output := c.String(flagOutput)
+	projectID := c.String(flagProject)
+
+	if err := validateOutputFormat(output); err != nil {
+		return err
+	}
+
+	client, err := getClient(c)
+	if err != nil {
+		return err
+	}
+
+	opts := meta.ListOptions{}
+
+	for {
+		secrets, err :=
+			client.Core().Projects().Secrets().List(c.Context, projectID, &opts)
+		if err != nil {
+			return err
+		}
+
+		switch strings.ToLower(output) {
+		case flagOutputTable:
+			table := uitable.New()
+			table.AddRow("KEY", "VALUE")
+			for _, secret := range secrets.Items {
+				table.AddRow(secret.Key, "*** REDACTED ***")
+			}
+			fmt.Println(table)
+
+		case flagOutputYAML:
+			yamlBytes, err := yaml.Marshal(secrets)
+			if err != nil {
+				return errors.Wrap(
+					err,
+					"error formatting output from get secrets operation",
+				)
+			}
+			fmt.Println(string(yamlBytes))
+
+		case flagOutputJSON:
+			prettyJSON, err := json.MarshalIndent(secrets, "", "  ")
+			if err != nil {
+				return errors.Wrap(
+					err,
+					"error formatting output from get secrets operation",
+				)
+			}
+			fmt.Println(string(prettyJSON))
+		}
+
+		if secrets.RemainingItemCount < 1 || secrets.Continue == "" {
+			break
+		}
+
+		// Exit after one page of output if this isn't a terminal
+		if !terminal.IsTerminal(int(os.Stdout.Fd())) {
+			break
+		}
+
+		if shouldContinue, err :=
+			shouldContinue(secrets.RemainingItemCount); err != nil {
+			return err
+		} else if !shouldContinue {
+			break
+		}
+
+		opts.Continue = secrets.Continue
+	}
+
+	return nil
+}
+
+func secretsSet(c *cli.Context) error {
+	projectID := c.String(flagProject)
+	kvPairsStr := c.StringSlice(flagSet)
+
+	// We'll make two passes-- we'll parse all the input into a map first,
+	// verifying as we go that the input looks good. Only after we know it's good
+	// will we iterate over the k/v pairs in the map to set secrets via the API.
```

Comment:
Would it make sense to check that the key and value conform to the schema, before sending it to the API?

Reply:
There are actually several spots where I debated whether it would be a good idea to do some client-side validation so that we could short-circuit and avoid a round-trip for something that was destined to fail, but I eventually came down on the side of keeping all validation server-side only because if you don't and you then need to relax or tighten the schema, you don't want to have to do that both client-side _and_ server-side-- or risk them disagreeing about what is valid. Opted to have the api server be the supremer arbiter of valid / invalid.

Reply:
I didn't consider that different clients could try enforcing different validation rules - in this case, this makes sense.